sled-agent/src/bootstrap/agent.rs

```diff
@@ -434,6 +540,66 @@ impl Agent {
         &self.baseboard
     }
 
+    async fn poll_ddmd_for_bootstore_peer_update(
+        log: &Logger,
+        bootstore_node_handle: bootstore::NodeHandle,
+    ) -> JoinHandle<()> {
+        let log = log.clone();
+        tokio::spawn(async move {
+            // Explicit fail fast. We can't do anything without a DdmAdminClient.
+            // Do we need to loop as in wicketd?
+            let ddmd_client = DdmAdminClient::localhost(&log).unwrap();
+            let mut current_peers: BTreeSet<SocketAddrV6> = BTreeSet::new();
+            loop {
+                // We're talking to a service's admin interface on localhost and
+                // we're only asking for its current state. We use a retry in a
+                // loop instead of `backoff`.
+                const RETRY: tokio::time::Duration =
+                    tokio::time::Duration::from_secs(5);
+
+                loop {
+                    match ddmd_client
+                        .derive_bootstrap_addrs_from_prefixes(&[
+                            BootstrapInterface::GlobalZone,
+                        ])
+                        .await
+                    {
+                        Ok(addrs) => {
+                            let peers: BTreeSet<_> = addrs
+                                .map(|ip| {
+                                    SocketAddrV6::new(ip, BOOTSTORE_PORT, 0, 0)
+                                })
+                                .collect();
+                            if peers != current_peers {
+                                current_peers = peers;
+                                if let Err(e) = bootstore_node_handle
+                                    .load_peer_addresses(current_peers.clone())
+                                    .await
+                                {
+                                    warn!(
+                                        log,
+                                        concat!("Bootstore comms error: {}. ",
+                                        "bootstrap agent must be terminating?.",
+                                    ),
```

Comment:
Why is this the case? I see you believe it because of the `return`, but I don't grok why any error from `load_peer_addresses` would imply this.

Reply:
You are right. I inferred this from the fact that `load_peer_addresses can only return a `NodeRequestError::Send` which can only result when the receiver side of the tokio MPSC channel hangs up. The receiver side is owned by the `bootstore::Node` which is running in its own task with a handle owned by the bootstrap agent. This task is supposed to run forever but of course it can panic due to programmer error driven invariant violations and possibly other things I'm unaware of. This means that this log message is incorrect.

However, I'm not really sure what to do in this scenario. The bootstore::Node is not supposed to stop. In theory it could be restarted, but it's probably best to panic here and let the OS process restart. What are your thoughts @smklein ?

Reply:
From the type-system, that's not super clear -- it's acting on a `NodeRequestError`, I think, which could have a variety of variants.

I'm not sure I'm even suggesting this, but  what would be wrong with "logging an error and retrying later?" If the bootstore is undergoing a transient error -- maybe it gets reset, like you said -- wouldn't we just want to loop again and refresh the set of peers at some point in the future?

(That also seems like it's justification for setting `current_peers = peers` only in the case where we successfully pass this info through the `bootstore_node_handle`).

Reply:
Here's the thing, I don't think the error here is transient. The `bootstore::Node` should never shutdown. The only reason it would shutdown is if it panic'd. Right now I only know of two explicit panics, and both of those are invariant violations due to programmer error that I want to know about immediately. I suppose we could respawn the task, but that doesn't make me very comfortable.

However, your point about the type system is spot on. I think for these types of permanent errors it would be nice to return a different type. The problem is that a tokio send error can happen on *any* API call, and is really unexpected. Instead of returning an error we could have the tokio task panic, but we never actually wait on the result, because we expect the task to run forever, as we do in other places. Honestly, I'm not really sure what to do here. 

Maybe the best thing is to do an explicit match on the `NodeRequestError` and panic if it's a `Send` variant, which it always will be in the case of this call.

Reply:
If we want to make a short-term move on this:

- I think it's okay for this task to exit on error, if it wants to, particularly if you're committing to the statement "if we see any error here, we should not retry".
- I would just be cautious about documenting "this is what an error means for the other side." It's this task's choice to exit, which is fine, but I just am scared of making inferences about what the error means from the other side.

Reply:
Ok, I think that's fair. We can log it and figure out why it crashed later. In the short term this will have practically no effect on the system as the bootstore is really only used during rack initialization. What will happen (until sled-agent is restarted one way or another) is that this sled will be unable to hand out shares to new sleds added to the rack and it will not be able to accept updates for its early networking information once we add that capability.

Reply:
Implemented in [d8fcf7e](https://github.com/oxidecomputer/omicron/commit/d8fcf7eecf302be91d647f228708259e2123fc23)